### ISSUE TITLE
Restoring from a backup

### ISSUE THREAD

**Issue by lewisdiamond** [digint/btrbk#152]
I had a little misshap, newegg shipped me RMA'd memory that corrupted my whole filesystem.

I wiped out my main filesystem and my local backup filesystem. I restored my external backup using `btrfs send | receive`.

This caused a whole lot of issues, for example:
```
WARNING: Receive target of "/media/btr0/snapshots/@root.20170504" exists at unexpected location: lede:/mnt/snapshots/@root.20170424
```

But more importantly, it created bad incremental backups. According to a btrfs developer, after restoring from a backup (sent out, received in), you can't do an incremental backup, it has to be a full backup.

The result of btrbk's incremental backups was not matching the backed up source subvolume and no error was thrown.

That was done on v0.24.0
btrfs-progs v4.10.2
Linux 4.10.11-1-ARCH

**Comment by digint** [comment]
> WARNING: Receive target of [...] exists at unexpected location: [...]

That's not a bug, btrbk tells you that the subvolume you are transmitting already exist on the target, but at a location not handled by btrbk (unexpected location). see also #145, #91

> But more importantly, it created bad incremental snapshots

This happens sometimes, I had other users experiencing this too, especially when databases are involved. (see #91)

> According to a btrfs developer, after restoring from a backup (sent out, received in), you can't do an incremental backup, it has to be a full backup.

I think this should work. Not sure what you did exactly, maybe you can explain in more detail?

**Comment by lewisdiamond** [comment]
Sorry I made a mistake when initially writing the bug report. It's not a bad incremental snapshot but rather a bad incremental backup. I updated the initial post.

Here's what happened:
I had a bunch of incremental backups on a remote target. I cleared all the local snapshots and backups, I `btrfs send | receive` the backup on a newly created btrfs fs then ran btrbk. The incremental backups are now still sent with -p to the remote location but btrfs requires a full backup at that point.

**Comment by lewisdiamond** [comment]
Re: WARNING: Receive target of [...] exists at unexpected location: [...]

The path and name of the snapshot matches exactly what's configured in the btrbk config. Those were backups made by BTRBK prior to the restore.

**Comment by lewisdiamond** [comment]
I may have found the issue. I'm waiting for a response from the btrfs mailing list but it seems like the Received UUIDs are not set properly, which may be causing the issues I see.

**Comment by digint** [comment]
> According to a btrfs developer, after restoring from a backup (sent out, received in), you can't do an incremental backup, it has to be a full backup.

I got a bit scared of this, so I ran a quick test to double-check. Conclusion: **incremental backup after restoring works**, at least for my simple unittest-setup.

@lewisdiamond can you provide me with a link documenting why this should not work? What I can think of is that it could fail on specific use cases, e.g. with some "nocow" files.

```
### setup test environment

mkdir /tmp/btrbk_unittest
mkdir /tmp/btrbk_unittest/mnt_source
mkdir /tmp/btrbk_unittest/mnt_target
truncate -s 100M /tmp/btrbk_unittest/btrfs0
truncate -s 100M /tmp/btrbk_unittest/btrfs1
losetup /dev/loop0 /tmp/btrbk_unittest/btrfs0
losetup /dev/loop1 /tmp/btrbk_unittest/btrfs1
mkfs.btrfs -L BTRBK_UNITTEST_SOURCE /dev/loop0
mkfs.btrfs -L BTRBK_UNITTEST_TARGET /dev/loop1
mount /dev/loop0 /tmp/btrbk_unittest/mnt_source
mount /dev/loop1 /tmp/btrbk_unittest/mnt_target

# create btrbk.conf
mkdir /tmp/btrbk_unittest/configs
cat > /tmp/btrbk_unittest/configs/btrbk.conf << EOF
snapshot_dir _btrbk_snap
snapshot_preserve_min all
target_preserve_min all
transaction_log /tmp/btrbk_unittest/tlog
volume /tmp/btrbk_unittest/mnt_source
subvolume svol
target send-receive /tmp/btrbk_unittest/mnt_target
EOF

# setup source subvolume (svol) for btrbk
btrfs subvolume create /tmp/btrbk_unittest/mnt_source/svol
mkdir /tmp/btrbk_unittest/mnt_source/_btrbk_snap
date > /tmp/btrbk_unittest/mnt_source/svol/CREATE.TAG


### create initial backup

btrbk -l debug -c /tmp/btrbk_unittest/configs/btrbk.conf run 2>>/tmp/btrbk_unittest/debug.log
# [...] output:
# +++ /tmp/btrbk_unittest/mnt_source/_btrbk_snap/svol.20170508
# *** /tmp/btrbk_unittest/mnt_target/svol.20170508


### add some data, and create incremental backup (2x)

date >> /tmp/btrbk_unittest/mnt_source/svol/MODIFY.TAG
dd if=/dev/urandom of=/tmp/btrbk_unittest/mnt_source/svol/rnd bs=1M count=1

btrbk -l debug -c /tmp/btrbk_unittest/configs/btrbk.conf run 2>>/tmp/btrbk_unittest/debug.log
# [...] output:
# +++ /tmp/btrbk_unittest/mnt_source/_btrbk_snap/svol.20170508_1
# >>> /tmp/btrbk_unittest/mnt_target/svol.20170508_1

date >> /tmp/btrbk_unittest/mnt_source/svol/MODIFY.TAG
dd if=/dev/urandom of=/tmp/btrbk_unittest/mnt_source/svol/rnd bs=1k count=1 seek=42k

btrbk -l debug -c /tmp/btrbk_unittest/configs/btrbk.conf run 2>>/tmp/btrbk_unittest/debug.log
# [...] output:
# +++ /tmp/btrbk_unittest/mnt_source/_btrbk_snap/svol.20170508_2
# >>> /tmp/btrbk_unittest/mnt_target/svol.20170508_2


### delete svol and snapshot (create scenario where restore is needed)

btrfs sub delete /tmp/btrbk_unittest/mnt_source/svol
btrfs sub delete /tmp/btrbk_unittest/mnt_source/_btrbk_snap/svol.20170508_1


### restore backup: mnt_target/svol.20170508_1 -> mnt_source/svol

mkdir /tmp/btrbk_unittest/mnt_source/restore
btrfs send -p /tmp/btrbk_unittest/mnt_target/svol.20170508 /tmp/btrbk_unittest/mnt_target/svol.20170508_1 | btrfs receive /tmp/btrbk_unittest/mnt_source/restore
btrfs sub snap /tmp/btrbk_unittest/mnt_source/restore/svol.20170508_1 /tmp/btrbk_unittest/mnt_source/svol


### backup is restored, run btrbk

btrbk -l debug -c /tmp/btrbk_unittest/configs/btrbk.conf run 2>>/tmp/btrbk_unittest/debug.log
# [...] output:
# +++ /tmp/btrbk_unittest/mnt_source/_btrbk_snap/svol.20170508_3
# >>> /tmp/btrbk_unittest/mnt_target/svol.20170508_3


### works fine, start modifying svol again

date >> /tmp/btrbk_unittest/mnt_source/svol/MODIFY.TAG
dd if=/dev/urandom of=/tmp/btrbk_unittest/mnt_source/svol/rnd bs=1k count=1 seek=12k

btrbk -l debug -c /tmp/btrbk_unittest/configs/btrbk.conf run 2>>/tmp/btrbk_unittest/debug.log
# [...] output:
# +++ /tmp/btrbk_unittest/mnt_source/_btrbk_snap/svol.20170508_4
# >>> /tmp/btrbk_unittest/mnt_target/svol.20170508_4
```

**Comment by digint** [comment]
> I may have found the issue. I'm waiting for a response from the btrfs mailing list but it seems like the Received UUIDs are not set properly, which may be causing the issues I see.

I checked your post on linux-btrfs: You are using `btrfs property set -ts /my/subvol ro false`, this confuses btrbk by leaving Received-UUID set.

**Comment by lewisdiamond** [comment]
Re: full backup after restore, I don't know the specifics but here's the relevant information
http://www.spinics.net/lists/linux-btrfs/msg44089.html

Re: rw -> ro, sure this may be what confuses btrfs but surely new snapshots should clear the Received UUID and that problem isn't only for btrbk, but btrfs send|receive itself fails because of that.

**Comment by digint** [comment]
> Re: full backup after restore, I don't know the specifics but here's the relevant information
http://www.spinics.net/lists/linux-btrfs/msg44089.html

Well this is a pretty old post... 
It says for send:
```
Send(m1(u1, #, #)*, m2) -> m2(u2, u1, -)*
```
which is not always true as of kernel 4.4, where things have changed (presumably to address exactly this issue). See this patch (which if I'm not mistaken is in the kernel since v4.4): https://www.mail-archive.com/stable@vger.kernel.org/msg141393.html

For **kernel>=4.4**, it is now (if source does not have received-uuid):
```
Send(m1(u1, -, #)*, m2) -> m2(u2, u1, -)*
```
and, if source has received-uuid="r1"
```
Send(m1(u1, r1, #)*, m2) -> m2(u2, r1, -)*
```
The same applies to send with parent (SendP).

This means that received-uuid stays intact between multiple send/receive operations, so that btrfs can identify "identical" subvolumes by matching received-uuid.

**Comment by lewisdiamond** [comment]
@digint I really don't know the details. Someone who knows way more about that stuff told me that on the btrfs irc channel. It may or may not be outdated information.

I ended up creating a full backup and moved to a new snapshot that wasnt `ro`, I still think that the way btrfs handles the Received UUID in that situation is wrong since it breaks send/receive.

The warnings that made no sense were also caused by this.

I'll mark this issue as closed, feel free to head to #btrfs if you would like to confirm incremental backups after restore works as you think (I hope so since it makes it much simpler).
